aqt/helper.py

```diff
@@ -103,13 +105,30 @@ def downloadBinaryFile(url: str, out: str, hash_algo: str, exp: str, timeout):
                         hash.update(chunk)
                     fd.flush()
             except Exception as e:
-                raise ArchiveDownloadError(f"Download error: {e}") from e
+                raise ArchiveDownloadError(f"Download of {filename} has error: {e}") from e
             if exp is not None and hash.digest() != exp:
-                raise ArchiveDownloadError(
-                    f"Download file is corrupted! Detect checksum error.\nExpected {exp}, Actual {hash.digest()}"
+                raise ArchiveChecksumError(
+                    f"Downloaded file {filename} is corrupted! Detect checksum error.\n"
+                    f"Expected {exp}, Actual {hash.digest()}"
                 )
 
 
+def retry_on_errors(action: Callable[[], any], acceptable_errors: Tuple, num_retries: int, name: str):
+    logger = getLogger("aqt.helper")
+    for i in range(num_retries):
+        try:
+            retry_msg = f": attempt #{1 + i}" if i > 0 else ""
+            logger.info(f"{name}{retry_msg}...")
+            ret = action()
+            if i > 0:
+                logger.info(f"Success on attempt #{1 + i}: {name}")
+            return ret
+        except acceptable_errors as e:
+            if i < num_retries - 1:
+                continue  # just try again
+            raise e from e
```

Comment:
This is the simplest way I could think of to write the "retry" function, while still logging some output when it's useful. I like keeping it separate from `downloadBinaryFile`, for readability and reusability, but we could move this logic directly into that function if you think that would be better.

The `acceptable_errors` parameter is a little weird, and very easy to misuse. It must contain a tuple of Exception classes, or the function will fail. I thought it was useful to pass the exceptions in this way; that's the only way this function can be reused effectively. On the other hand, this is such a strange thing to do, I'm concerned that this could be an antipattern.